### ISSUE TITLE
Idea: drop Index file in table with very short partitions

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#13312]
A user complained that a table with a huge number of very short partitions was surprisingly big - perhaps as much as 4 times larger than if the same data is stored in a modest number of large partitions.

Let's look at a simple example. Consider two tables, each of them has three integer columns, `p, c, x`:

* In table1, `p` is the partition key, `c`  is the clustering key, `x` is a regular column.
* In table2, `(p,c)` is a compound partition key, `x` is a regular column.

We write a million rows to each table, with `(p, c, x) = (1, i, 1)` for one million i's.
So both tables have exactly the same data - table1 is one partition with one million rows, and table2 in one million partitions, but both have exactly the same rows.

It turns out that after compaction, the size of table1's sstables is 8.2 MB, and the size of table2 is 34 MB. table2 is more than 4 times larger than table1! To understand why, let's look at the size of the individual sstable components:

1. For table1, almost the entire 8.2 MB size is the "Data" component. The "Index" component is almost empty (just one partition).
2. For table2, the "Data" component is 12.6 MB, the "Index" component is 20MB. The "Filter" is 1.2 MB.

It is not surprising that table2's Data component is slightly larger than table1's (12.6 MB vs 8.2 MB) - after all the individual partitions do have some overhead (e.g., a tombstone), and this overhead is noticable when the partitions are so tiny (just a single integer). It's also not surprising that the Bloom filter (the "Filter" file) takes more space when we have many partitions. But what is really surprising, and really frustrating is the size of the Index file, which is almost twice bigger than the Data file, which we didn't expect.

The idea I want to propose in this issue is that when partitions are very short, it would be better not to have an Index file at all. The Summary file could point directly into the Data file instead of the Index file. 

I don't know what should be the threshold for dropping the Index file - for larger partitions, it may still be good. Maybe we can write the Data and Index file as we do today, and after-the-fact, if we notice that Index is larger than Data (or even if it is larger than half of Data), we delete the Index component and rewrite the Summary.

Instead of dropping the Index file, and even more efficient thing to do can be to create in Index a level between the Data and Summary - in other words, Index will be a sample of Data's keys - not every paritition as in Data but not a sparse a sample as Summary, but something in the middle. But this will require more work to implement.

This use case, of very short partitions, may seem artificial, but a real user encountered it with a materialized view - the user had a base table with reasonably-long partitions, but then had a view with a compound partitions, where each row of data was in its own partition - and each row was also very short. The user didn't even realize that this very-short-partitions case was happening, but was surprised that the view was 4 times larger than the base table.

Code for the tests described above:
```python
@pytest.fixture(scope="function")
def table1(cql, test_keyspace):
    t = f'{test_keyspace}.{unique_name()}'
    cql.execute(f'CREATE TABLE {t}(p int, c int, x int, PRIMARY KEY (p, c))')
    yield t
    cql.execute(f'DROP TABLE {t}')

@pytest.fixture(scope="function")
def table2(cql, test_keyspace):
    t = f'{test_keyspace}.{unique_name()}'
    cql.execute(f'CREATE TABLE {t}(p int, c int, x int, PRIMARY KEY ((p, c)))')
    yield t
    cql.execute(f'DROP TABLE {t}')

# Table with a single 1-million-row partition, each row has, in addition
# to the key, a single int.
def test_single_partition(cql, table1):
    table = table1
    stmt = cql.prepare(f"INSERT INTO {table} (p, c, x) VALUES (1, ?, 1)")
    for i in range(1000000):
        cql.execute(stmt, [i])
        if (i%100000)==0:
            print(i)
            nodetool.flush(cql, table)
            nodetool.compact(cql, table)
    nodetool.flush(cql, table)
    nodetool.compact(cql, table)
    print('going to sleep\n')
    time.sleep(10000)

# Table with a single 1-million partitions, each with a single row
# (no clustering key), and each row has additionally an int value
def test_million_partitions(cql, table2):
    table = table2
    stmt = cql.prepare(f"INSERT INTO {table} (p, c, x) VALUES (1, ?, 1)")
    for i in range(1000000):
        cql.execute(stmt, [i])
        if (i%100000)==0:
            print(i)
            nodetool.flush(cql, table)
            nodetool.compact(cql, table)
    nodetool.flush(cql, table)
    nodetool.compact(cql, table)
    print('going to sleep\n')
    time.sleep(10000)
```

**Comment by michoecho** [comment]
I agree with the underlying idea (that it's better to index with sector/page granularity instead of partition granularity), but I disagree with the implementation suggestion. One day you are "dropping the Index file" because two index levels is too much and the index uses too much disk, another day you will be "adding a SummarySummary file" because two index levels is not enough and the Summary doesn't fit in RAM anymore. If we are going to do major changes to the index format, we might as well design a sane arbitrary-level tree instead of replacing the existing (not sane) hard-coded two-level tree with a hard-coded two-or-one-level tree.

**Comment by nyh** [comment]
@michoecho I agree that the whole three-level Summary->Index->Data structure could be replaced by a more a general tree structure, but as you said, this would be a major change and I don't see us doing it anytime soon. Also, in most setups, the benefit of replacing the three-level tree by a more general tree will be very minor. 

Instead, in this issue I'm looking at one **extreme** situation where a table has only tiny partitions, and the index file more than doubles the total storage. Although this extreme situation can sound silly (who ever wants a table with only one integer column per partition?), it was actually seen by a real user - who created a materialized view with just one integer column.

I agree that other extreme situations might require other hacks. We already have an issue for the summary-file-being-too-big (#3147). But I think the number of such hacks will be minimal - my thinking is that we won't use these "hacks" to save 10% of disk space or 10% of memory or 10% of runtime here and there, but only in extreme situation which can yield at least 2-fold improvement of something.

**Comment by avikivity** [comment]
I don't see the problem with the index file, just reduce its granularity so that each entry covers multiple partitions. Not sure that's legal though in terms of what readers expect.

**Comment by michoecho** [comment]
> I agree that the whole three-level Summary->Index->Data structure could be replaced by a more a general tree structure, but as you said, this would be a major change and I don't see us doing it anytime soon.

@nyh Why? Do you think it would be too risky (in terms of regression potential) or too time-consuming for what it's worth? I think you are underestimating the relative difficulty of the change you are proposing vs a general solution.

> Instead, in this issue I'm looking at one **extreme** situation

> I agree that other extreme situations might require other hacks.

> But I think the number of such hacks will be minimal

But why go that way at all? The good old B-tree would handle *all* cases well. Why do we instead have a design which works about as well as a B-tree in the average case but breaks down in extreme cases?

Usually when you choose the mathematically worse data structure it's either because it's simpler or because constant factors make it perform better in practice. But that's not the case here. As far as I can see, the current design is in every way worse than the obvious design.

It's not only about size either. Do you realize that a read from a table filled with 10-byte partitions is a (decimal) order of magnitude *more* expensive (both in terms of CPU usage and disk bandwidth usage) than a read from a table (of equal cardinality!) filled with 200-byte partitions? That's because the fixed design of the index makes the size of index pages inversely proportional to the size of partitions (more precisely: to the `key size : partition size` ratio).

For very small partitions, the index pages are so large that 95% of resources is spent reading and parsing index pages. It's terrible.

**Comment by michoecho** [comment]
> The idea I want to propose in this issue is that when partitions are very short, it would be better not to have an Index file at all. The Summary file could point directly into the Data file instead of the Index file.
> 
> I don't know what should be the threshold for dropping the Index file - for larger partitions, it may still be good. Maybe we can write the Data and Index file as we do today, and after-the-fact, if we notice that Index is larger than Data (or even if it is larger than half of Data), we delete the Index component and rewrite the Summary.

@nyh This is a very bad idea because Summary can't index clustering keys. You can end up have such a mix of big and small partitions that the index file ends up bigger than the data file, but you will have some large partitions in the middle. If you delete the index and use only the summary, you might end up having to do 1GiB reads to get a 10 B row.

> I don't see the problem with the index file, just reduce its granularity so that each entry covers multiple partitions. Not sure that's legal though in terms of what readers expect.

@avikivity As-is? Definitely not legal — it would require some non-trivial reader changes.

**Comment by avikivity** [comment]
> 
> @avikivity As-is? Definitely not legal — it would require some non-trivial reader changes. (Edit: okay, that depends on the definition of "trivial" — the changes shouldn't be big either.)

We can't change the reader in older versions of Scylla or in Cassandra, so it would be a format change.

If we change the format, might as well make it a b-tree. There's also the upcoming [CEP-25](https://cwiki.apache.org/confluence/display/CASSANDRA/CEP-25%3A+Trie-indexed+SSTable+format).

**Comment by nyh** [comment]
I've renamed this issue to be about "Wasteful storage of table with very short partitions" - not about the specific solution I proposed which @avikivity and @michoecho didn't like.  We may end up closing this issue when we eventually (?) do a complete redesign of the sstable indexing components (Summary and Index).

**Comment by nyh** [comment]
By the way, the very-short-partitions case in the above example had a 1.2 MB Bloom filter, which is small compared to the 32 MB of other files (Data and Index), but remember that the Bloom filter resides entirely in memory, so it's bad that it's very big. In this case the Bloom filter's false-positive rate should be increased so the file will be smaller. See #2024.

**Comment by sagar-keshri** [comment]
Hi Team,
Is there any update on this issue? 
We are facing this issue in multiple clusters

**Comment by mykaul** [comment]
@bhalevy - let's consider this for 6.1. Unsure who should own it.

**Comment by avikivity** [comment]
We can consider implementing Cassandra's BTI. It's using Trie instead of BTree, but with various compression options that make it similar to btree.

We should evaluate it on interesting data layouts (like this one here), and on the amount of memory needed to generate it (i.e. whether it can be streamed or if we need to hold large amounts of indexing information in memory). Maybe it matters less with tablets since sstable sizes have tighter bounds.
